segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/BaseChunkStorageProvider.java

```diff
@@ -0,0 +1,592 @@
+
+
+    protected static final OpStatsLogger READ_LATENCY = STATS_LOGGER.createStats(MetricsNames.STORAGE_READ_LATENCY);
+    protected static final OpStatsLogger WRITE_LATENCY = STATS_LOGGER.createStats(MetricsNames.STORAGE_WRITE_LATENCY);
+    protected static final OpStatsLogger CREATE_LATENCY = STATS_LOGGER.createStats(MetricsNames.STORAGE_CREATE_LATENCY);
+    protected static final OpStatsLogger DELETE_LATENCY = STATS_LOGGER.createStats(MetricsNames.STORAGE_DELETE_LATENCY);
+    protected static final OpStatsLogger CONCAT_LATENCY = STATS_LOGGER.createStats(MetricsNames.STORAGE_CONCAT_LATENCY);
+
+    protected static final Counter READ_BYTES = STATS_LOGGER.createCounter(MetricsNames.STORAGE_READ_BYTES);
+    protected static final Counter WRITE_BYTES = STATS_LOGGER.createCounter(MetricsNames.STORAGE_WRITE_BYTES);
+    protected static final Counter CONCAT_BYTES = STATS_LOGGER.createCounter(MetricsNames.STORAGE_CONCAT_BYTES);
+
+    protected static final Counter CREATE_COUNT = STATS_LOGGER.createCounter(MetricsNames.STORAGE_CREATE_COUNT);
+    protected static final Counter DELETE_COUNT = STATS_LOGGER.createCounter(MetricsNames.STORAGE_DELETE_COUNT);
+    protected static final Counter CONCAT_COUNT = STATS_LOGGER.createCounter(MetricsNames.STORAGE_CONCAT_COUNT);
+
+    protected static final Counter LARGE_CONCAT_COUNT = STATS_LOGGER.createCounter(MetricsNames.STORAGE_LARGE_CONCAT_COUNT);
+
+    protected final AtomicBoolean closed;
+
+    /**
+     * Constructor.
+     *
+     */
+    public BaseChunkStorageProvider() {
+        this.closed = new AtomicBoolean(false);
+    }
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports truncate operation on underlying storage object.
+     *
+     * @return True or false.
+     */
+    @Override
+    abstract public boolean supportsTruncation();
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports append operation on underlying storage object.
+     *
+     * @return True or false.
+     */
+    @Override
+    abstract public boolean supportsAppend();
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports merge operation either natively or through appends.
+     *
+     * @return True or false.
+     */
+    @Override
+    abstract public boolean supportsConcat();
+
+    /**
+     * Gets a value indicating whether this Storage implementation supports native merge operation on underlying storage object.
+     *
+     * @return True or false.
+     */
+    @Override
+    abstract public boolean supportsNativeConcat();
+
+
+    /**
+     * Determines whether named file/object exists in underlying storage.
+     *
+     * @param chunkName Name of the storage object to check.
+     * @return True if the object exists, false otherwise.
+     */
+    @Override
+    public boolean exists(String chunkName)  throws IOException {
```

Comment:
Is this method supposed to be overridden?

Reply:
The derived classes should not override the  methods from ChunkStorageProvider interface. They are supposed to override all the abstract methods with naming pattern doXYZ()  ..(line 406) .

Reply:
@eolivelli I have split a really big PR in multiple parts. 
Part 2 contains actual implementation for the ChunkStorageProvider.  
Below are the links to the implementation. They all override the abstract methods.  
- HDFS https://github.com/sachin-j-joshi/pravega/blob/issue-4676-PDP-34-incremental-implementation-2/bindings/src/main/java/io/pravega/storage/hdfs/HDFSChunkStorageProvider.java

- NFS https://github.com/sachin-j-joshi/pravega/blob/issue-4676-PDP-34-incremental-implementation-2/bindings/src/main/java/io/pravega/storage/filesystem/FileSystemChunkStorageProvider.java

- ECS https://github.com/sachin-j-joshi/pravega/blob/issue-4676-PDP-34-incremental-implementation-2/bindings/src/main/java/io/pravega/storage/extendeds3/ExtendedS3ChunkStorageProvider.java

Reply:
I suggest to make these methods "final", this way it is not possible to override them

Reply:
Fixed. Thanks